IN DEVELOPMENT Crossfire rssi and Link Quality from the receiver protocol link statistics

laz2wiringpi opened the pull request:
Crossfire   rssi and Link Quality from the receiver protocol link statistics 
When crsf is selected from the profile and rssi Source is none then it will switch to rssi_sorce of RSSI_SOURCE_RX_PROTOCOL_CRFS
 
LINK Quality  osd  0-300  
 
Rssi  percent  0-100

rssi dbm 0-130 in osd
"osd_rssi_dbm_alarm" 0-300
 
new items in osd settings  for  "link quality" and "rssi dbm" in osd->alarm
 
3 debug_mode settings 
"CRSF_LINK_STATISTICS_UPLINK",
     Uplink RSSI Ant. 1 ( dBm * -1 )
     Uplink RSSI Ant. 2 ( dBm * -1 )
     Uplink Package success rate / Link quality ( % )
     RF Mode (  4fps = 0 , 50fps =1, 150hz =2)
"CRSF_LINK_STATISTICS_PWR",
     Diversity active antenna ( enum ant. 1 = 0, ant. 2 )
     Uplink SNR ( db )
     Uplink TX Power ( enum 0mW = 0, 10mW, 25 mW, 100 mW, 500 mW, 1000 mW, 2000mW ) ** 
 
"CRSF_LINK_STATISTICS_DOWN",
     Downlink RSSI ( dBm * -1 )
     Downlink package success rate / Link quality ( % )
     Downlink SNR ( db )
* Uplink is the connection from the ground to the UAV and downlink the opposite direction.
** enum for mico tx  not in documentation  0mw = 0 , 25mw,100mw, 250mw 
 #8001  
 #5901

laz2wiringpi commented:
@codecae PR for crossfire posted thanks for the help.

codecae commented:
Looking better!   Could be further optimized, though.  Will peer review again in a few.

codecae commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +116,107 @@ struct crsfPayloadRcChannelsPacked_s {
+    uint8_t Downlink_Link_quality;
+    uint8_t Downlink_SNR;
+
+}crsfPayloadLinkstatistics_t;
```
space before the type definition and newlines around struct members can be removed

laz2wiringpi replied:
fixed

laz2wiringpi commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +116,107 @@ struct crsfPayloadRcChannelsPacked_s {
+        setRssiDirect(0,RSSI_SOURCE_RX_PROTOCOL_CRFS);
+        setLinkQualityDirect(0);
+
+    }
```
the check is for missing packets 
if i handle it in the packet i will  not know its missing

codecae replied:
I see what you're doing here.  I misunderstood your approach.  You're validating the timeliness of the rssi frames upon receipt of the channel frames.  Disregard.  Keeping `LastLinkStatisticsFrameUs` static in this approach makes sense.   The logic could also then be removed from the handler.  Sorry for the misunderstanding.

codecae replied:
Only other suggestion I would have would be 
```
static void crsfCheckRssi(timeUs_t currentTimeUs)
{
```

laz2wiringpi replied:
no problem  ..... 
in my tests i found when i turned of transmitter would keep  rssi  but would lower ql to 40-20 ish 
static got it ...

laz2wiringpi commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +116,107 @@ struct crsfPayloadRcChannelsPacked_s {
+
+    setLinkQualityDirect( (crsfRXLinkstatistics.RF_Mode == 2) ? (100 + crsfRXLinkstatistics.Uplink_Link_quality) : crsfRXLinkstatistics.Uplink_Link_quality);
+
+}
```
i get the idea but the code has issues 
lastTimeUs = currentTimeUs;  is called twice .. 
and 
static timeUs_t lastTimeUs = currentTimeUs; 
 error: initializer element is not constant

ill work on it

codecae replied:
It's just a suggestion so it could have issues.  Just trying to share the spirit of the concept is all.  you could declare it without initialization and the second cycle should pick up normally.

codecae replied:
Cleaned it up a bit.. is this better?  This assumes you're keeping `static timeUs_t lastLinkStatisticsFrameUs` outside of the handler

```
static void handleCrsfLinkStatisticsFrame(const crsfLinkStatistics_t* statsPtr, timeUs_t currentTimeUs)
{
    const crsfLinkStatistics_t stats = *statsPtr;
    lastLinkStatisticsFrameUs = currentTimeUs;
    switch (debugMode) {
    case DEBUG_CRSF_LINK_STATISTICS_UPLINK:
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 0, stats.Uplink_RSSI_1);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 1, stats.Uplink_RSSI_2);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 2, stats.Uplink_Link_quality);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 3, stats.RF_Mode);
        break;
    case DEBUG_CRSF_LINK_STATISTICS_PWR:
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_PWR, 0, stats.active_antenna);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_PWR, 1, stats.Uplink_SNR);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_PWR, 2, stats.Uplink_TX_Power);
        break;
    case DEBUG_CRSF_LINK_STATISTICS_DOWN:
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_DOWN, 0, stats.Downlink_RSSI);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_DOWN, 1, stats.Downlink_Link_quality);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_DOWN, 2, stats.Downlink_SNR);
        break;
    }
    const uint8_t rssidBm = stats.active_antenna ? stats.Uplink_RSSI_2 : stats.Uplink_RSSI_1;
    const uint8_t rssidBmScaled = scaleRange(crsfConvertRssidBm(rssidbm), 1, 100, 0, RSSI_MAX_VALUE);
    const uint8_t linkQuality = stats.Uplink_Link_quality + ((stats.RF_Mode == 2) ? 100 : 0);
    setRssiDirect(rssidBmScaled, RSSI_SOURCE_RX_PROTOCOL_CRSF);
    setLinkQualityDirect(linkQuality);
}
```

laz2wiringpi replied:
i cleaned it up a bit 
your help is greatly appreciated. 
 
need to retest it all . and verify it sitll works

laz2wiringpi commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +116,107 @@ struct crsfPayloadRcChannelsPacked_s {
+}crsfPayloadLinkstatistics_t;
+
+static crsfPayloadLinkstatistics_t crsfRXLinkstatistics;
+static uint32_t LastLinkStatisticsFrameUs;
```
ok you told me this before i moved the functions from telemetry. before the pr . missed it in the move 
fixed

codecae replied:
Nit-pick alert `timeUs_t`

laz2wiringpi commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +116,107 @@ struct crsfPayloadRcChannelsPacked_s {
+static crsfPayloadLinkstatistics_t crsfRXLinkstatistics;
+static uint32_t LastLinkStatisticsFrameUs;
+
+int converdbm( int x){
```
will the compiler handle is we pass 8 bits and not 16 to the function ?
why should it be a int16_t and not int8_t?

codecae replied:
int8_t should be sufficient

laz2wiringpi commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +116,107 @@ struct crsfPayloadRcChannelsPacked_s {
+    if (x > -21){
+        return 100;
+    }
+    return  (int)(((-0.0154*x*x)-(0.3794*x)+98.182) + 0.5f);
```
yep your right  uint8_t here

laz2wiringpi commented on src/main/rx/crsf.c:
```diff
@@ -178,6 +281,23 @@ STATIC_UNIT_TESTED void crsfDataReceive(uint16_t c, void *data)
+                             if (
+                                 (rssiSource == RSSI_SOURCE_RX_PROTOCOL_CRFS) &&
+                                 (crsfFrame.frame.deviceAddress == CRSF_ADDRESS_FLIGHT_CONTROLLER) &&
+                                 (crsfFrame.frame.frameLength == CRSF_FRAME_ORIGIN_DEST_SIZE + CRSF_FRAME_LINK_STATISTICS_PAYLOAD_SIZE)
```
RSSI_SOURCE_RX_PROTOCOL_CRFS is set if source is none in init. and if you set it for rc channel 
it will not write

laz2wiringpi replied:
ok if they make it longer thats fine .. but if the shorten it ... ouch... 
and what if the expand one to 16 better not to work than to start doing strange things.
i think its better to stop working  than read unknown data.
     (crsfFrame.frame.frameLength == CRSF_FRAME_ORIGIN_DEST_SIZE + CRSF_FRAME_LINK_STATISTICS_PAYLOAD_SIZE)

laz2wiringpi replied:
(crsfFrame.frame.deviceAddress == CRSF_ADDRESS_FLIGHT_CONTROLLER) &&
 saw some other packets not to CRSF_ADDRESS_FLIGHT_CONTROLLER  ie  the data was the same like an echo. 
may be a bug on there side . im have to run beta firmware for another issue ... 
no point processing other's data ..

codecae replied:
Ya that one is truly debatable.  Not sure how many folks will have a network of devices communicating link quality to each other but I suppose it's possible.  Better safe than sorry.

laz2wiringpi commented on src/main/rx/crsf.c:
```diff
@@ -192,6 +312,8 @@ STATIC_UNIT_TESTED uint8_t crsfFrameStatus(rxRuntimeConfig_t *rxRuntimeConfig)
 {
     UNUSED(rxRuntimeConfig);
 
+    crsfcheckrssi(micros());
```
need it to set rssi link to 0 when it stops giving packets over 10hz

laz2wiringpi commented on src/main/rx/rx.c:
```diff
@@ -365,16 +365,30 @@ static uint8_t updateLinkQualitySamples(uint8_t value)
 {
+
 #ifdef USE_RX_LINK_QUALITY_INFO
+    if (rssiSource != RSSI_SOURCE_RX_PROTOCOL_CRFS) {
```
ok i had to look 2 or 3 times .. dam. what was i drinking...
ill clean it up ..

codecae replied:
hah!  no worries :)  easy to miss.

laz2wiringpi commented on src/main/rx/rx.c:
```diff
@@ -365,16 +365,30 @@ static uint8_t updateLinkQualitySamples(uint8_t value)
+    #ifdef USE_RX_LINK_QUALITY_INFO
+
+        linkQuality = linkqualityValue;
+
```
yep.

laz2wiringpi commented:
@codecae  
ill squash after i confirm it still works . 
but the magic smoke just cam out of my test board .. 😒

laz2wiringpi commented on src/main/rx/crsf.c:
```diff
@@ -161,59 +159,52 @@ int converdbm( int x){
-       }
-
+    uint8_t rssidBmScaled = 0;
+    uint8_t linkQuality = 0;
```
uint8_t rssidBmScaled = 0;
    uint8_t linkQuality = 0;
need to remove .

codecae commented:
> but the magic smoke just cam out of my test board .. 😒
Oh no!  Short?

codecae commented on src/main/rx/crsf.c:
```diff
@@ -146,13 +145,12 @@ typedef struct crsfPayloadLinkstatistics_s {
 
-int converdbm( int x){
+static uint8_t crsfConvertRssidBm(int8_t x) {
     x = x * -1;
```
nitpick alert
`x *= -1`

laz2wiringpi replied:
remove need to convert from -1
static uint8_t crsfConvertRssidBm(uint8_t x)
{
    if (x > 92){
      return 1;
    }
    if (x < 21){
      return 100;
    }
    return (uint8_t)(((-0.0154*x*x)-(-0.3794*x)+98.182) + 0.5f);
}

codecae commented on src/main/rx/crsf.c:
```diff
@@ -134,8 +134,7 @@ typedef struct crsfPayloadRcChannelsPacked_s crsfPayloadRcChannelsPacked_t;
 
-typedef struct crsfPayloadLinkstatistics_s {
-
+typedef struct crsfPayloadLinkstatistics_s{
```
If I don't say this, someone else will:
`typedef struct crsfPayloadLinkStatistics_s {`

codecae commented on src/main/rx/crsf.c:
```diff
@@ -146,13 +145,12 @@ typedef struct crsfPayloadLinkstatistics_s {
     uint8_t Downlink_RSSI;
     uint8_t Downlink_Link_quality;
     uint8_t Downlink_SNR;
+}crsfLinkStatistics_t;
```
If I don't say this, someone else will:
`} crsfPayloadLinkStatistics_t;`

codecae replied:
From this, you could also
`#define CRSF_FRAME_LINK_STATISTICS_PAYLOAD_SIZE (sizeof(crsfLinkStatistics_t))`

laz2wiringpi replied:
CRSF_FRAME_LINK_STATISTICS_PAYLOAD_SIZE was already in crsf_protocal.h

laz2wiringpi replied:
fixed sp(s)

codecae commented on src/main/rx/crsf.c:
```diff
@@ -146,13 +145,12 @@ typedef struct crsfPayloadLinkstatistics_s {
-}crsfPayloadLinkstatistics_t;
-
-static crsfPayloadLinkstatistics_t crsfRXLinkstatistics;
+//static crsfPayloadLinkstatistics_t crsfRXLinkstatistics;
```
delete

laz2wiringpi replied:
yep

codecae commented:
All-in-all, looking much cleaner (imho)... on the home stretch.  Feeling better to you?

laz2wiringpi commented:
not happy rssi is not working in % debug's dbm matches 
i also am getting mode 1,2,1,2,......1,2  more than expected !
bit again its at 10mw .

need to test more   🤔

laz2wiringpi commented:
@codecae 
ok for another  round .
not real happy with rssi calc dbm to % . but so far its the best i can come up with. 

im happy with LQ output . tried to use the sample average nope must be direct. do to mode changes.

codecae commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +116,100 @@ struct crsfPayloadRcChannelsPacked_s {
+    uint8_t Uplink_RSSI_1;
+    uint8_t Uplink_RSSI_2;
+    uint8_t Uplink_Link_quality;
+    uint8_t Uplink_SNR;
```
based on the spec:
`int8_t Uplink_SNR`

laz2wiringpi replied:
your right SNR can be +-  
will fix !

codecae commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +116,100 @@ struct crsfPayloadRcChannelsPacked_s {
+    uint8_t Uplink_TX_Power;
+    uint8_t Downlink_RSSI;
+    uint8_t Downlink_Link_quality;
+    uint8_t Downlink_SNR;
```
`int8_t Downlink_SNR`

laz2wiringpi replied:
> I'm looking at the RSSI conversion. If I understand correctly, the frame is providing dBm *-1 and you want to convert this to %?

sack discussion ?

codecae commented:
I'm looking at the RSSI conversion.  If I understand correctly, the frame is providing dBm *-1 and you want to convert this to %?

laz2wiringpi commented:
@codecae > I'm looking at the RSSI conversion. If I understand correctly, the frame is providing dBm *-1 and you want to convert this to %?

yea , ideally i would like it to be like other's in bf but the dbm upper and lower values are 12-130db in the manual but that does not seem to be reality .

i may just debug_mode them in some test flight's and see if i can come up with something more realistic.

slack conversation?

codecae commented:
Based on the crossfire documentation, the nominal range is -1 to -130dbm ("good" to "bad") .. why not just scale that range and call it a day?

laz2wiringpi commented:
> Based on the crossfire documentation, the nominal range is -1 to -130dbm ("good" to "bad") .. why not just scale that range linearly to from 100 to 1 (respectively) and call it a day?

did that it was bad ...  look at rssi in the radio. yuck..  🤢

and it states in the manual as low as -12 .. yea right! never see below 23 if im right next to the quad 
10 feet 30 ect. not sure in real flight .  never looked rssi  in crsf.
you konw the  issue's will be my rssi never gets above xx  👀

going flying...

codecae commented:
This I think is precisely why TBS doesn't recommend the use of RSSI to gauge signal strength and suggests the use of LQ. It is what it is at the end of the day.  To me, it would be better to implement it purely as dBm and create a new widget for SNR (or combination of the two in a single widget).  I think, in context, this is more valuable to long range pilots than a % RSSI.  (just my opinion I guess)

laz2wiringpi commented:
> This I think is precisely why TBS doesn't recommend the use of RSSI to gauge signal strength and suggests the use of LQ. It is what it is at the end of the day.< 

 To me, it would be better to implement it purely as dBm and create a new widget for SNR (or combination of the two in a single widget). I think, in context, this is more valuable to long range pilots than a % RSSI. (just my opinion I guess)

baby steps ...  
   just going to get % as close as i can for now.  Ive thought about remove it for this pr any opinion? 

Other  Pr(s) . 
1. document for crossfire and BF. how you up for that? 
2. Downlink SRN  warnings.  this s a major goal for me. i have areas that are just un-flyable do rf noise
would be nice to know before im in mode 1-0 . happen to me  today. !
3. Rssi dbm  and SNR for long range. not into  a long range. (no where for me to do so in may area,)
4. a way to set power like vtx on arm disarm 
   That needs  Team Black Sheep CRSF protocol #6917  .. i could not figure out the unit test. for a re base.

codecae commented:
To me, a baby step would be to implement it just as it comes across in dBm.  This is exactly what is being displayed on the OLED screen when using the large crossfire,so it would at least be consistent. Otherwise, perhaps @TBSChin and team could lend an ear and help us find out what the actual calculation should be from the manufacturer's perspective.

laz2wiringpi commented:
> To me, a baby step would be to implement it just as it comes across in dBm. This is exactly what is being displayed on the OLED screen when using the large crossfire,so it would at least be consistent. Otherwise, perhaps @TBSChin and team could lend an ear and help us find out what the actual calculation should be from the manufacturer's perspective.

The issue is BF is set up as %0-99 for alarms ect. 

The more i think about the dbm mess the more i think Bf needs another Rssi by Dbm and not %
but that's beyond the scope i had planed for this pr . 

I should just do Quality link drop the rssi  and leave that for another Pr.   IMO most uses will not use RSSI and just want link quality. 

any objections ?

bizmar commented:
From my (users) perspective i would really really appreciate 0-300 LQ so I could know what mode i'm in. That is the most important part to me, as a race and freestyle pilot who wants consistent 150hz. A new element that just says 0,1,2 would be fine too.

(for now I have it "hacked" together with openTX)

codecae commented:
0-300 would change the widget from 8 to 16 bit... Not a big deal, but has other possible implications

laz2wiringpi commented:
> 0-300 would change the widget from 8 to 16 bit... Not a big deal, but has other possible implications

yep 
 keeping 199-0 where 199 to 100 is mode 2,  99-1 is mode 1, and if your in mode 0 its 0.
this is compatible with other code and the other pr for the alarm for link quality

codecae commented:
How are you feeling about this one?  Getting there?

laz2wiringpi commented:
Ive cherry picked to my 4.01 , fixing my test quad for a weekend test flights.

laz2wiringpi commented:
@codecae 
changed 
osd alarms/stats only log 99-1

Rssi 0-100 dbm
add setting "osd_rssi_dbm = on" alarms/stats by dbm

can you re-check, i think its good to go . going to test on my other quads.

laz2wiringpi commented:
@mikeller i know your quite over loaded right now. if you could take a look any feed back would be appreciated. 
also 
 is  #6917  dead letter now ?  may  provide source api ?

mikeller commented:
@laz2wiringpi: Can do - I normally hold off on work marked 'in development', but if you think it is about to be ready for feedback I am happy to review.

I don't see #6917 as anything like 'dead' - my hope is still that TBS will open up CRSF so that it can be used as a means to configure Betaflight. #6917 implements a good approach to integrating this into Betaflight, and will still be needed and valuable once TBS has published the library defining their side of the API.

laz2wiringpi commented:
> @laz2wiringpi: Can do - I normally hold off on work marked 'in development', but if you think it is about to be ready for feedback I am happy to review.
  just worried i overlooked something and crash do to a  mistake in memory. testing on my non junk quads ... next weekend 
or there is a better way to handle x y or z. 

> I don't see #6917 as anything like 'dead' - my hope is still that TBS will open up CRSF so that it can be used as a means to configure Betaflight. #6917 implements a good approach to integrating this into Betaflight, and will still be needed and valuable once TBS has published the library defining their side of the API.
good to hear ... i would like to set power level on arm  at some point like the vtx.

mikeller commented on src/main/osd/osd_elements.c:
```diff
@@ -1119,11 +1122,19 @@ static void osdElementWarnings(osdElementParms_t *element)
+              osdFormatMessage(element->buff, OSD_FORMAT_MESSAGE_BUFFER_SIZE, "RSSI LOW");
+              SET_BLINK(OSD_WARNINGS);
+              return;
+          }
```
This repetition is a bit inelegant - I'd probably go
```
if (osdWarnGetState(OSD_WARNING_RSSI) {
    bool warningIsOn = osdConfig()->osd_rssi_dbm ? getRssiPercent() > osdConfig()->rssi_alarm : getRssiPercent() < osdConfig()->rssi_alarm;
    if (warningIsOn) {
```

mikeller commented on src/main/osd/osd_elements.c:
```diff
@@ -1536,10 +1547,18 @@ void osdUpdateAlarms(void)
-        SET_BLINK(OSD_RSSI_VALUE);
-    } else {
-        CLR_BLINK(OSD_RSSI_VALUE);
+    if (osdConfig()->osd_rssi_dbm){
```
Same as above.

mikeller commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +117,90 @@ struct crsfPayloadRcChannelsPacked_s {
+    int8_t Downlink_SNR;
+} crsfLinkStatistics_t;
+
+static uint32_t LastLinkStatisticsFrameUs;
```
Lower case.

codecae replied:
`timeUs_t` ?

laz2wiringpi replied:
ops!  codece ... pointed  this out before. i must have not have not staged the fix and lost the changes   
cleaning it up.

laz2wiringpi replied:
changed timeUs_t LastLinkStatisticsFrameUs

mikeller commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +117,90 @@ struct crsfPayloadRcChannelsPacked_s {
+    uint8_t linkQuality = 0;
+    if ((stats.RF_Mode >= 1) && (stats.Uplink_Link_quality != 0)){
+        linkQuality = stats.Uplink_Link_quality -1;
+        if (stats.RF_Mode == 2 ){
```
Missing space.

laz2wiringpi replied:
fixed

mikeller commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +117,90 @@ struct crsfPayloadRcChannelsPacked_s {
+    if ((stats.RF_Mode >= 1) && (stats.Uplink_Link_quality != 0)){
+        linkQuality = stats.Uplink_Link_quality -1;
+        if (stats.RF_Mode == 2 ){
+            linkQuality = linkQuality +100;
```
Missing space.

laz2wiringpi replied:
fixed

mikeller commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +117,90 @@ struct crsfPayloadRcChannelsPacked_s {
+}
+#endif
+
+static void crsfcheckrssi(uint32_t currentTimeUs){
```
CamelCase. Missing space.

laz2wiringpi replied:
crsfCheckRssi removed space

codecae commented on src/main/rx/crsf.c:
```diff
@@ -114,6 +117,90 @@ struct crsfPayloadRcChannelsPacked_s {
+    setRssiDirect(rssidBmScaled, RSSI_SOURCE_RX_PROTOCOL_CRSF);
+
+    uint8_t linkQuality = 0;
+    if ((stats.RF_Mode >= 1) && (stats.Uplink_Link_quality != 0)){
```
`if (stats.RF_Mode && stats.Uplink_Link_quality)`

laz2wiringpi replied:
no longer relevent 
setLinkQualityDirect((stats.rf_Mode * 100) + stats.uplink_Link_quality);

laz2wiringpi commented:
> Looks mostly good. There are a bunch of cosmetic problems, and quite a bit of missing spaces. Please look through your code and fix missing spaces.

sorry i thought i had cleaned then up  having a hard time with my ide atom. it seems to format stuff and i cant seem to stop it . i think i made it worse somehow. CamelCase is a old dog ... hard to break the habit.
 
> One thing I am not sure about is 'abusing' `getRssiPercent()` to carry a dBm value - this will at least require renaming of the getter. But I think that even this is not quite what we want - a lot of places like telemetry protocols are using `getRssiPercent()`, and most TX systems are hard wired to interpret the RSSI value as 'less is worse', and not as the other way round (which is incorrect in the first place, as 'more dBm' also means 'more power' (https://www.rapidtables.com/convert/power/dBm_to_mW.html)). So what we should do is have `getRssiPercent()` return an approximated percentage value, and add `getRssiDbm()` to return a dBm value.

yea... abusing good term for the mess ...   

however i lost you after that point.
1. tried a conversion from  dbm to percentage. no luck . 
2. tried just dbm . would like the alarms to work.

is your suggestion  to getRssiDbm and  not use getRssiPercent ? that may work . 

  getRssiPercent just will not work.  i spent hours attempting but as soon as i change power levels or area  changes   .. no luck .
i even asked Tbs .  that did not go so well.  there stuck in there interpretation of how it should be.

what i have learned is that the raw dbm value is what is  for long distance flyers.  im not and no area to do so unless i'm over water! not my idea of a relaxing flight.

codecae commented:
RSSI percentage isn't really a good indicator of signal strength like it is for other systems.  What is low RSSI for some systems is considered nominal in crossfire.  dBm and SNR are really better suited for long distance and LQ is better for close to mid range.  What did TBS say to do?

mikeller commented:
@laz2wiringpi, @codecae: Even if RSSI percentage is not a good system for signal strength in the case at hand, it is what is used by everything that consumes the RSSI data. Just replacing it with a measure that works inverse of it will break a lot of things, so I don't think this is a good way forward.
Even if we resort to returning a completely made up value as RSSI value, like `100 - rssiDbm`, this is still more likely to work for users than returning `rssiDbm` for something where consumers expect 'lower is worse'.

codecae commented:
Completely agree.  I mentioned once before that we should just linearly scale and constrain a percentage based on how TBS has documented a nominal range in the crossfire manual.

> RSSI - Received signal strength indicator, measurement of the power present in a received radio
signal [typ. -1dBm (good) to -130dBm (bad), logarithmic scale]

Default to 1-130 and perhaps give users the ability to change the boundaries.

codecae commented:
Sorry I bumped the button on accident 😬

laz2wiringpi commented:
ill agree to disagree... 
1. I dont think i seen a request for rssi just link quality. 

2. in the manual it states something like at 50db twice the distance should be 56db .. this may be true but i have not seen that . in a vacuum ? using a YAGI  ?  

3. while i don't wish to abuse the code as i did  mike did give a great pointer on how it may be done .  so  working on the direction of getRssiDbm() and his elegant code witch i really like by the way !  bool warningIsOn = osdConfig()->osd_rssi_dbm ... getRssiDbm() ?  getRssiPercent()

 4. for the 100-1 the is an option some where i saw rsssi_inverted. 

 5. At some point 
 A. give up and do the quick an dirty rssi as codex stated many times.  
   or B. have a solution that will never please everyone that ist's just the way of a dev ...we all know that... right!.  

 6.  Crossfire wikki   

Were a long way off from 4.1 ..  So i have time to tinker . feeling a lot better with mike's code suggestion .  

"measure that works inverse of it will break a lot of things," 
this however does worry me ..?
 if i return a conversion for  getRssiPercent() and have a new  getRssiDbm() ... should be safe.

mikeller commented:
@laz2wiringpi: Good point about 'time to tinker' - better to get something good later than something that doesn't work for some cases sooner.

@codecae:

> RSSI - Received signal strength indicator, measurement of the power present in a received radio signal [typ. -1dBm (good) to -130dBm (bad), logarithmic scale]

Good input. This would suggest that `(130 - rssiDbm) / 1.3f` should give us something that is very close to what other protocols return as 'RSSI percentage'. If we feed this into RSSI processing early enough, users will be able the RSSI scaling to map this to the range that they want.

In addition to that, I would still like to see a `int16_t getRssiDbm()` function to return the 'pure' RSSI dBm value, so that it can be used in places like OSD. Probably go for a separate element though, so users get the choice of having calculated RSSI (in the range / direction they are used to) or direct dBm, or both.

laz2wiringpi commented:
"addition to that, I would still like to see a int16_t getRssiDbm() function to return the 'pure' RSSI dBm value, so that it can be used in places like OSD."

@mikeller that's the plan .. 
 
 "Probably go for a separate element though, or direct dBm, or both."
.... osd is over loaded thought i heard something say that.
 one of the  reason's i liked your code less cycles to do the work. 

 so i thought i would use the same element so one one or the other . not both. ... dont see the need . with all this stuff   you wont be able to see where your going :)

dont know .. easy to debug with  two elements though. that's aplus..

mikeller commented:
@laz2wiringpi: Using both elements concurrently is more of an edge case. Allowing users to either use the 'traditional' RSSI, or dBm values, and doing so in a clean way that does not need extra parameters is the motivation behind using separate elements.

laz2wiringpi commented:
@mikeller @codecae this more like what were aiming for?
fixed up spaces, added rssidbi - one todo need to figure out the default alarm value 70 ...

laz2wiringpi commented:
I need to recheck, test and cleanup here is the progress so far.
 
link Quality 0-300 uint16_t
rssidbm 0-130
updated /added alarms, osd edit's  for the above 
CRSF_LINK_STATISTICS

laz2wiringpi commented:
I have decided after a discussion with etracer and jeyper that I will be closing this pr so I can split it into new planed pr(s).
1.	CRSF_LINK_STATISTICS
2.	Link Quality will also allow the use of a second Rssi percent for those using redundancy transmitters.
3.	Rssi by dbm 0-130 lower is better.

And will  use this a reference of the suggestions.  
I would like to thank you for you help and guidance.

laz2wiringpi commented:
ok i split it out now to work on the details... its the way to go.

TonyBlit commented:
I'm a little late to the party, but here's my two cents regarding the specific case of CRSF which is what I use/know.

Showing the LQ on the existing separate widget is definitely the way to go, but I have some comments regarding RSSI.

Let me start saying that in my opinion LQ alone is useless for the vast majority of users, and RSSI/LQ is way too cryptic. I've seen several reports of people flying about 1km away with an LQ of 99 when it suddenly drops (due to hitting a null antenna orientation or an interference). LQ as a signal indicator degrades very abruptly, so if you only rely on it chances are you're missing the whole picture. I think LQ is only complementary/extra information to RSSI and it was not the best idea on TBS side to offer RSSI/LQ or LQ alone.

TBS does not explain how they calculate RSSI (even when asked directly by an iNav developer, although they suggested him to use RSNR for his purposes in the LUA script). I did a bit of testing myself and RSS, RSNR and RSSI looked to be highly linearly correlated to each other. But other people with more experience working with LoRa suggest that RSS is a more reliable indicator when SNR is high, and SNR itself is a better indicator on low SNR scenarios.

TBS recommends different warning/critical values for RSSI and LQ, but since RSSI is not available as telemetry data, for setting warnings on your TX they end up just saying "use LQ". As explained above, this is a bad idea and this PR is an opportunity to fix this.

What I would do is to expose RSSI as a new telemetry value (don't confuse with 1RSS/2RSS which by the way are always sending the same value even in a diversity receiver), and ask people to stop configuring RSSI as RSSI/LQ or LQ in the RX (eventually/hopefully TBS might remove this feature from the firmware as well). This way, we will be able to set alarms in our TX based on RSSI with the TBS calculated value. And also RSSI will be stored in the TX logs (which currently isn't).

Calculating RSSI out of RSS and/or RSNR will also be interesting so you will be able to save a channel that is currently wasted with the RSSI mapping. I have a few ideas to find a formula as close as possible to TBS', but they involve a bit of "crowdsourcing" :-)

For the time being, as I don't care about high RSS scenarios, I'm currently using RSNR as a trigger for my TX alarms. LoRa can theoretically work with up to -12dB of SNR, so I've set up an alarm when going below 0 which means entering the "magic zone" of communicating below noise floor. According to my logs, this should be better for my own purposes than just emitting warnings based on LQ. Will let you know how it goes in a few months. :-P

Note: since I'm covering a few topics I wasn't sure where to put this comment in the split PRs.

laz2wiringpi commented:
@TonyBlit thanks for the info ..
cant read it all now ...  will look at it latter i've been looking for someone that may have some more knowledge on the subject .. FYI ... 

"don't confuse with 1RSS/2RSS which by the way are always sending the same value even in a diversity receiver"
  what i get ... two different   values ... may be a  TBS bug got  fixed?

TonyBlit commented:
I'm seeing the same readings on 1RSS and 2RSS on a Diversity Nano RX with firmware 2.93. What RX and firmware version are you using?

laz2wiringpi commented:
Sorry internet down its documented that fw has issues w that rx 
I when I get on slack
